sway-core/src/asm_generation/checks.rs

```diff
@@ -74,3 +77,52 @@ fn check_for_contract_opcodes(ops: &[AllocatedOp]) -> CompileResult<()> {
         err(vec![], errors)
     }
 }
+
+/// Checks for a violation of the Checks-Effects-Interaction pattern.
+/// A check could be anywhere, but what we want to ensure is that effects happen before interactions. An interaction is a `CALL` opcode, and an effect is a `SWW` or `SWWQ` opcode.
+///
+///If a `CALL` opcode happens _before_ a `SWW`/`SWWQ` opcode _without_ any jumps away, then we know this is an interaction before an effect.
+///
+///
+///For every opcode:
+///  if this is a `SWW`/`SWWQ`:
+///     if seen_call:
+///       throw CEI violation
+///
+///  if this is a jump (non-conditional):
+///     seen_call = false;
+///  if this is a CALL:
+///     seen_call = true;
+///
+/// TODO: We need to do this at the storage API level, or improve the span here. The span often
+/// just points to the standard library that is doing the storing.
+fn check_for_cei_violation(ops: &[AllocatedOp]) -> CompileResult<()> {
+    let mut warnings = vec![];
+    let mut seen_call = false;
+    use AllocatedOpcode::*;
+    let default_span = sway_types::span::Span {
+        span: pest::Span::new("no span found for opcode".into(), 0, 1).unwrap(),
+        path: None,
+    };
+    for op in ops {
+        match op.opcode {
+            SWW(..) | SWWQ(..) => {
+                if seen_call {
+                    warnings.push(CompileWarning {
+                        span: op
+                            .owning_span
+                            .clone()
+                            .unwrap_or_else(|| default_span.clone()),
+                        warning_content: Warning::CEIViolation,
+                    });
+                }
+            }
+            CALL(..) => seen_call = true,
+            JI(..) | RET(..) | RVRT(..) => {
+                seen_call = false;
```

Comment:
Does a `JI` always jump outside of this list of ops? Or can it jump to another location within this list?

Reply:
I feel like `JI` could just anywhere (i.e. an `if` statement), so it shouldn't be used to clean whether a call was seen?

Reply:
^ That's essentially coming with IR. A more SSA graph-based representation.

 You guys are right, a `JI` should only clear the flag if it jumps to _after_ the potentially offending state write.